app/main.py

```diff
@@ -1,44 +1,87 @@
-    toilet_id = service.register_device(toilet)
 
-    # create a few programs
-    wake_up_program = [
-        Message(hue_light_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.PLAY_SONG, "Rick Astley - Never Gonna Give You Up"),
-    ]
+    # gather registering devices simultaneously
+    hue_light_id, speaker_id, toilet_id = await asyncio.gather(
+        service.register_device(hue_light),
+        service.register_device(speaker),
+        service.register_device(toilet),
+    )
 
-    sleep_program = [
-        Message(hue_light_id, MessageType.SWITCH_OFF),
-        Message(speaker_id, MessageType.SWITCH_OFF),
-        Message(toilet_id, MessageType.FLUSH),
-        Message(toilet_id, MessageType.CLEAN),
-    ]
+    # combine run_sequence and run_parallel for wake_up_program
+    await run_sequence(
+        run_parallel(
+            service.send_msg(Message(hue_light_id, MessageType.SWITCH_ON)),
+            service.send_msg(Message(speaker_id, MessageType.SWITCH_ON)),
+        ),
+        run_parallel(
+            service.send_msg(Message(hue_light_id, MessageType.SWITCH_ON)),
+            service.send_msg(
+                Message(speaker_id, MessageType.PLAY_SONG, "Rick Astley - Never Gonna Give You Up")
+            ),
+        ),
+    )
 
-    # run the programs
-    service.run_program(wake_up_program)
-    service.run_program(sleep_program)
+    # combine run_sequence and run_parallel for sleep_program
+    await run_sequence(
+        run_parallel(
+            service.send_msg(Message(hue_light_id, MessageType.SWITCH_OFF)),
+            service.send_msg(Message(speaker_id, MessageType.SWITCH_OFF)),
+        ),
+        service.send_msg(Message(toilet_id, MessageType.CLEAN)),
+        service.send_msg(Message(toilet_id, MessageType.FLUSH)),
+    )
+
+    print("=====END OF PROGRAM======")
+
+    await run_sequence(
+        service.unregister_device(hue_light_id),
+        service.unregister_device(speaker_id),
+        service.unregister_device(toilet_id),
+    )
+
+
+async def run_sequence(*functions: asyncio.Task) -> None:
+    tasks = [asyncio.create_task(func) for func in functions]
+    for task in tasks:
+        await task
+        if task.get_name() == "send_msg":
+            device_id = task.get_coro().cr_frame.f_locals["msg"].device_id
+            device = service.get_device(device_id)
+            print(f"{device.__class__.__name__} received message.")
+            await device.send_message(MessageType(task.get_coro().cr_frame.f_locals["msg"].msg_type))
+
+        if task.get_name() == "unregister_device":
+            device_id = task.get_coro().cr_frame.f_locals["device_id"]
+            device = service.get_device(device_id)
+            print(f"Disconnecting {device.__class__.__name__}.")
+            await device.disconnect()
+            await asyncio.sleep(TIME_TO_SLEEP)
+            print(f"{device.__class__.__name__} disconnected.")
```

Comment:
So much not needed code....
just do:
```
    for function in functions:
        await function
```

Reply:
done, but the performing time has twice as increased!
<img width="1660" alt="Screenshot 2023-07-03 at 14 47 56" src="https://github.com/mate-academy/py-asyncio-iot/assets/10571459/ae895be6-0eb9-414f-a9b6-1fbdc9220256">

Reply:
[Solution5_even faster](https://github.com/mate-academy/py-asyncio-iot/pull/122/commits/be5832b96f8ca6356e627f4805335f351e95b0a0)
<img width="967" alt="Screenshot 2023-07-03 at 16 44 44" src="https://github.com/mate-academy/py-asyncio-iot/assets/10571459/cdbbd044-d705-44f0-8b4d-e1545602b1f4">